judoscale-good_job/lib/judoscale/good_job/metrics_collector.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+require "judoscale/job_metrics_collector"
+require "judoscale/job_metrics_collector/active_record_helper"
+require "judoscale/metric"
+
+module Judoscale
+  module GoodJob
+    class MetricsCollector < Judoscale::JobMetricsCollector
+      def self.adapter_config
+        Judoscale::Config.instance.good_job
+      end
+
+      def collect
+        metrics = []
+        time = Time.now.utc
+
+        all_queues = ::GoodJob::JobsFilter.new({}).queues.keys
+        self.queues |= all_queues
+
+        # TODO: silence query logs for this
+        t = ::GoodJob::Job.arel_table
+        run_at_by_queue = ::GoodJob::JobsFilter.new(state: "queued").filtered_query.order(
+          :queue_name, t.coalesce(t[:scheduled_at], t[:created_at])
+        ).pluck(
+          Arel.sql("distinct on (queue_name) queue_name, #{t.coalesce(t[:scheduled_at], t[:created_at]).to_sql}")
+        ).to_h
```

Comment:
I'd feel better if this didn't use `GoodJob::JobsFilter`; I don't want to guarantee the durability of the JobsFilter's interface.

- `COALESCE(scheduled_at, created_at)` is the correct way to determine when a job _should_ begin performing.
- `performed_at` will give you when the job starts performing. (If a job performance is interrupted by a SIGKILL, that value will be inaccurate until the job is re-performed, but that probably is an edge case). 

Therefore, the query for max queuing latency per queue would something like this:

```SQL
  SELECT queue_name, max(NOW() - COALESCE(scheduled_at, created_at))
  FROM good_jobs
  WHERE 
    performed_at IS NULL
    AND finished_at IS NULL -- unnecessary maybe
  GROUP BY queue_name
```   

I think that SQL is pretty durable across GoodJob versions. I don't mind if you wanted to use Active Record on `GoodJob::Executions` either (`GoodJob::Executions` is 1:1 to the `good_jobs` table which has a separate row for every retry, whereas `GoodJob::Job` has a default scope to filter out the retries)

Reply:
> My understanding is that it filters out the retries that have already failed, right?

Yes and I don't think it's necessary to add on the additional scopes of the `Job` model to generate a general-purpose metric. `GoodJob::Job` is not used operationally in GoodJob, only to make building the dashboard easier because it matches the human display model.

> I bring this up because I was thinking the GoodJob::Job.queued scope could be useful here.

That scope looks at Advisory Locks which isn't performant. It's to handle the edge case I mentioned of SIGKILL for Dashboard Display, but _operationally_ I wouldn't recommend using it. 

I think the Delayed Job metrics collector is closer to what I'm imagining would be most durable in that it just runs a SQL query against the table: 

https://github.com/judoscale/judoscale-ruby/blob/23c9059ddbe47b7cc4df1d889e0396f4fdded190/judoscale-delayed_job/lib/judoscale/delayed_job/metrics_collector.rb#L20-L27

I think doing it with Active Record / Arel to get some things for free (e.g. database connection and table names) is good, but I don't consider those scopes stable.